### PR TITLE
add Module for managing user in InfluxDB

### PR DESCRIPTION
**Pull request by zhhuta**
##### SUMMARY
Add module to manage users in InfluxDB

##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request

##### COMPONENT NAME
influxdb_user

```lib/ansible/modules/database/influxdb/influxdb_user.py```

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = /Users/zhhuta/.ansible.cfg
  configured module search path = [u'/Users/zhhuta/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.10 (default, Jul 15 2017, 17:16:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+version_added: 2.5
+author: "Vitaliy Zhhuta (@zhhuta)"
+requirements:
+    - "python >= 2.6"
+    - "influxdb >= 0.9"
+options:
+    hostname:
+        description:
+            - The hostname or IP address on which InfluxDB server is listening
+        required: false
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: there is a missing doc here
~~~diff
+        default: localhost
~~~

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    username:
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: in other influxdb modules, this is the user to authenticate, so please streamline with the existing modules related to influxdb
(I agree login_username is a much better name but we can not change existing modules without breaking users playbooks)

Please change this to `user_name` and adjust python code as well.

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        description:
+            - User that we want to create
+        default: None
+        required: True
+    password:
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: please change this to `user_password` for prevent a conflict with existing influxdb modules

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        description:
+            - Password that we want to set for user
+        default: None
+        required: false
+    admin:
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py:42]: while we change params names: what about `admin_role`?

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            - The port on which InfluxDB server is listening
+        default: 8086
+        required: false
+    state:
+        description:
+            - Determines if the database should be created or destroyed
+        choices: ['present', 'absent']
+        default: present
+        required: false
+    login_username:
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: streamline with existing modules: `username`

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        description:
+            - user to auth wirh influxdb
+        default: root
+        required: false
+    login_password:
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: streamline with existing modules: `password`

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+def influxdb_argument_spec():
+    return dict(
+        hostname=dict(default='localhost', type='str'),
+        port=dict(default=8086, type='int'),
+        login_username=dict(default='root', type='str'),
+        login_password=dict(default='root', type='str', no_log=True),
+        username=dict(required=True, type='str'),
+        password=dict(required=False, type='str', no_log=True),
+        admin=dict(default='False', type='str')
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: make this a boolean 
~~~diff
-        admin=dict(default='False, type='str')
+        admin=dict(default=False, type='bool')
~~~

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    )
+
+
+def connect_to_influxdb(module):
```
Comment (resmo) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: note there is a module_utils for the common code https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/influxdb.py (see code existing influxdb modules how to integrate).

**Review by resmo — CHANGES_REQUESTED** [review]
tested code ad worked as expected, however, the module needs to streamline params with exisiting influxdb modules

File: lib/ansible/modules/database/influxdb/influxdb_user.py (hunk 4 above)
Comment (zhhuta) [reply lib/ansible/modules/database/influxdb/influxdb_user.py:42]: changed

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        try:
+            client.drop_user(user_name)
+        except exceptions.InfluxDBClientError as e:
+            module.fail_json(msg=e.content)
+
+    module.exit_json(changed=True)
+
+
+def main():
+    argument_spec = influxdb_argument_spec()
```
Comment (Akasurde) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: Please use existing "influxdb_argument_spec" from https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/influxdb.py#L22

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        description:
+            - user to auth wirh influxdb
+        default: root
+        required: false
+    password:
+        description:
+            - user to auth wirh influxdb
+        default: root
+        required: false
+
```
Comment (Akasurde) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: Use existing documentation fragment - `extends_documentation_fragment: influxdb`

File: lib/ansible/modules/database/influxdb/influxdb_user.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        default: present
+        required: false
+    username:
+        description:
+            - user to auth wirh influxdb
+        default: root
+        required: false
+    password:
+        description:
+            - user to auth wirh influxdb
```
Comment (Akasurde) [inline lib/ansible/modules/database/influxdb/influxdb_user.py]: Password to authenticate with influxdb

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@zhhuta There is another PR open for this same feature - #20272

**Comment by zhhuta** [comment]
as soon as changes to https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/influxdb.py 
has been applied I'm going to remove from code 
       database_name=dict(default='None', type='str')

**Comment by resmo** [comment]
the database_name has been removed from utils, please remove your workarounds and rebase to latest.

**Comment by zhhuta** [comment]
@resmo actually we don't need to remove simply add in influxdb_argument_spec() database_name=dict(defult='None', type='str') -  we still have compatibility with others modules and if someone don't use it is will have None

File: lib/ansible/module_utils/influxdb.py [hunk 12]
```diff
@@ -44,6 +44,7 @@ def influxdb_argument_spec():
             port=dict(default=8086, type='int'),
             username=dict(default='root', type='str'),
             password=dict(default='root', type='str', no_log=True),
+            database_name=dict(defult='None', type='str'),
```
Comment (resmo) [inline lib/ansible/module_utils/influxdb.py]: Please remove this as it is not common across all influxdb modules

**Review by resmo — CHANGES_REQUESTED** [review]
only one thing left, then it is good to go

**Review by zhhuta — COMMENTED** [review]
all request are completed

**Review by resmo — APPROVED** [review]

**Review by Akasurde — APPROVED** [review]

**Comment by resmo** [comment]
shipit

**Comment by Akasurde** [comment]
@zhhuta Thanks for your contribution. @resmo Thanks for review.

**Comment by zhhuta** [comment]
@Akasurde @resmo  thank you too, will continue to contribute

**Comment by zhhuta** [comment]
shipit

**Pull request merged**